### PR TITLE
ANN activation functions for NMODL: `sigmoid(x)`, `relu(x)` and `tanh(x)`

### PR DESCRIPTION
**Pull request by llandsmeer**
I have been playing around a bit with getting Neural ODE's working (and in general ANN's as black-box homeostatic mechanisms) and found arbor lacking in ANN activation functions. Now not everybody might agree that this is actually needed in Arbor, so I don't expect this PR to be merged directly. But if people like the idea, here it is :)

The PR contains 3 common unary functions used in ANN's as nonlinear activations: `sigmoid(x)`, `relu(x)` and `tanh(x)`

Here is a very minimal example of a 'neural ode':

```
NEURON {
  SUFFIX node
}

PARAMETER {
    A1 = -0.662 A2 =  0.282 A3 =  0.957
    A4 = -0.189 A5 = -1.794 A6 =  1.094
    A7 = -1.133 A8 =  0.625 A9 =  0.074
    b1 = -1.434 b2 = -0.358 b3 = -0.093
}

STATE {
    x
    y
    z
}

INITIAL {
    x = 0
    y = 1
    z = 1
}

DERIVATIVE dstate {
    x' = tanh(A1 * x + A2 * y + A3 * z + b1)
    y' = tanh(A4 * x + A5 * y + A6 * z + b2)
    z' = tanh(A7 * x + A8 * y + A9 * z + b3)
}

BREAKPOINT {
    SOLVE dstate METHOD sparse
}
```

### CODE REVIEW

File: doc/fileformat/nmodl.rst [hunk 1]
```diff
@@ -173,8 +173,11 @@ Arbor-specific features
   step(x)             heaviside step with half value         :math:`\begin{align*} 1 & ~~ \text{if} ~x \gt 0, \\ 0 & ~~ \text{if} ~x \lt 0, \\ 0.5 & ~~ \text{otherwise}. \end{align*}`
   signum(x)           sign of argument                       :math:`\begin{align*} +1 & ~~ \text{if} ~x \gt 0, \\ -1 & ~~ \text{if} ~x \lt 0, \\ 0 & ~~ \text{otherwise}. \end{align*}`
   exprelr(x)          guarded exponential                    :math:`x e^{1-x}`
+  sigmoid(x)          sigmoidal function                     :math:`1/(1+e^{-x})`
+  relu(x)             rectified linear function              :math:`max(0, 1)
```
Comment (boeschf) [inline doc/fileformat/nmodl.rst:177]: I think the mathematical definition for the `relu` is wrong in the documentation.

Comment (llandsmeer) [reply doc/fileformat/nmodl.rst:177]: Good catch

**Comment by boeschf** [comment]
Small nitpick: could you add
``` 
        "using ::std::tanh;\n"
```
[here](https://github.com/llandsmeer/arbor/blob/nmodl-ai/modcc/printer/cprinter.cpp#L176)?

File: doc/dev/simd_api.rst [hunk 2]
```diff
@@ -594,6 +594,18 @@ In the following:
...
+      - *S*
+      - Lane-wise :math:`x \mapsto tanh(x)`
+
+    * - ``relu(s)``
+      - *S*
+      - Lane-wise :math:`x \mapsto relu(x)`
+
+    * - ``sigmoid(s)``
+      - *S*
+      - Lane-wise :math:`x \mapsto sigmoid(x)`
```
Comment (boeschf) [inline doc/dev/simd_api.rst:607]: maybe it is worth repeating the definition of those functions here

File: modcc/symdiff.cpp [hunk 3]
```diff
@@ -265,6 +265,48 @@ class SymPDiffVisitor: public Visitor, public error_stack {
         result_ = make_expression<IntegerExpression>(loc, 0);
     }
 
+    void visit(TanHUnaryExpression* e) override {
+        // 1 - tanh(f(x))^2 * df(x)/dx
```
Comment (boeschf) [inline modcc/symdiff.cpp:269]:
```suggestion
        // (1 - tanh(f(x))^2) * df(x)/dx
```

**Comment by boeschf** [comment]
can you add specializations for constants here: 
https://github.com/llandsmeer/arbor/blob/bfe83e80e52aebdf3a52075febc061bfae81bc44/modcc/symdiff.cpp#L437?

**Comment by boeschf** [comment]
performance metrics are not recorded for the new functions, see
https://github.com/llandsmeer/arbor/blob/nmodl-ai/modcc/perfvisitor.hpp
Not sure how to qualify `tanh` - maybe warrants its own category?

**Comment by llandsmeer** [comment]
Haven't yet found where the modcc tests are defined..

**Comment by boeschf** [comment]
tests are in `test/unit-modcc`, if that's what you were asking? The rest looks good, but the performance metrics are still missing

**Comment by thorstenhater** [comment]
Comment about the names: Is there a 'standard' sigmoid? I thought `tanh` is considered one as well on grounds 
of being S-shaped?

**Comment by llandsmeer** [comment]
TanH is definitely a sigmoidal function. But in ANN's when you specify `sigmoid` you mean the `1/(1 + exp(-x))` thing

Eg:

 - [Theano](https://theano-pymc-test.readthedocs.io/en/readthedocs/library/tensor/nnet/nnet.html#theano.tensor.nnet.nnet.sigmoid)
 - [Tensorflow](https://www.tensorflow.org/api_docs/python/tf/math/sigmoid)
 - [PyTorch](https://pytorch.org/docs/stable/generated/torch.nn.Sigmoid.html)
 - [Matlab](https://www.mathworks.com/help/deeplearning/ref/dlarray.sigmoid.html)
 - [ONNX](https://github.com/onnx/onnx/blob/main/docs/Operators.md#Sigmoid)

all define `sigmoid(x) = 1 / (1 + exp(-x))`. Although I've never seen an official declaration about this. It's probably mostly convention

**Comment by thorstenhater** [comment]
Ok, that's fine with me as long as it's clearly stated.

**Review by boeschf — APPROVED** [review]
